File: CSharp/CSharp/Game.cs
using System;
using System.Collections.Generic;

namespace CSharp
{
    public class Game
    {
        public const int MapWidth = 5;
        public const int MapHeight = 5;
        public static readonly HashSet<int> RulesStayAlive = new HashSet<int>(new[] { 2, 3, 5 });
        public static readonly HashSet<int> RulesBecomeAlive = new HashSet<int>(new[] { 0, 3 });

        public bool[,] Map { get; private set; }

        public Game()
        {
            Map = new bool[MapWidth, MapHeight];
        }

        public void PlayStep()
        {
            var temp = new bool[MapWidth, MapHeight];
            Array.Copy(Map, temp, Map.Length);

            for (int x = 0; x < MapWidth; x++)
            {
                for (int y = 0; y < MapHeight; y++)
                {
                    var currentCell = temp[x, y];
                    var neighborsCount = GetLivingNeighbors(temp, x, y);

                    if (currentCell && !RulesStayAlive.Contains(neighborsCount))
                    {
                        Map[x, y] = false;
                    }
                    else if (!currentCell && RulesBecomeAlive.Contains(neighborsCount))
                    {
                        Map[x, y] = true;
                    }
                }
            }
        }

        private static int GetLivingNeighbors(bool[,] map, int x, int y)
        {
            int count = 0;

            // Top-Left
            if (x > 0 && y > 0 && map[x-1, y-1])
            {
                count++;
            }

            // Top
            if (y > 0 && map[x, y - 1])
            {
                count++;
            }

            // Top-Right
            if (x < MapWidth - 1 && y > 0 && map[x + 1, y - 1])
            {
                count++;
            }

            // Left
            if (x > 0 && map[x - 1, y])
            {
                count++;
            }

            // Right
            if (x < MapWidth - 1 && map[x + 1, y])
            {
                count++;
            }

            // Bottom-Left
            if (x > 0 && y < MapHeight - 1 && map[x - 1, y + 1])
            {
                count++;
            }

            // Bottom
            if (y < MapHeight - 1 && map[x, y + 1])
            {
                count++;
            }

            // Bottom-Right
            if (x < MapWidth - 1 && y < MapHeight - 1 && map[x + 1, y + 1])
            {
                count++;
            }

            return count;
        }
    }
}

File: FSharp/Game.fs
namespace FSharp

module Game =
    type CellState = | ALIVE | DEAD
    
    let mapWidth = 50
    let mapHeight = 50
    let rulesBecomeAlive = [ 0; 3 ]
    let rulesStayAlive = [ 2; 3; 5 ]
    
    let toInt (map: CellState array) index =
        if (index >= 0 && index < map.Length) then
            match map.[index] with
            | ALIVE -> 1
            | DEAD -> 0
        else 0
        
    let private getLivingNeighbors map index =
        let toInt = toInt map
        let topLeft = index - mapWidth - 1 |> toInt
        let top = index - mapWidth |> toInt
        let topRight = index - mapWidth + 1 |> toInt
        let left = index - 1 |> toInt
        let right = index + 1 |> toInt
        let bottomLeft = index + mapWidth - 1 |> toInt
        let bottom = index + mapWidth |> toInt
        let bottomRight = index + mapWidth + 1 |> toInt      
        
        topLeft + top + topRight + left + right + bottomLeft + bottom + bottomRight
        
    let private checkRules rules neighborsCount =
        if (List.contains neighborsCount rules) then ALIVE else DEAD
        
    let private checkBecomeAliveRules = checkRules rulesBecomeAlive
    let private checkStayAliveRules = checkRules rulesStayAlive
        
    let private getNewState map index currentState =
        let neighborsCount = getLivingNeighbors map index
        match currentState with
        | ALIVE -> checkStayAliveRules neighborsCount
        | DEAD -> checkBecomeAliveRules neighborsCount
    
    let createEmptyMap() =
        Array.init (mapWidth * mapHeight) (fun i -> DEAD)
        
    let computeNextStep map =
        let getNewStateForMap = getNewState map            
        map |> Array.mapi getNewStateForMap
        
        
    createEmptyMap () |> computeNextStep |> printfn "%A";;